zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/factory/SharedZkClientFactory.java

```diff
@@ -117,6 +117,7 @@ private ZkConnectionManager getOrCreateZkConnectionManager(
   private void cleanupConnectionManager(ZkConnectionManager zkConnectionManager) {
     synchronized (_connectionManagerPool) {
       zkConnectionManager.close(true);
```

Comment:
When skipIfWatched == true, the close() call may not actually disconnect the zkConnectionManager given shared watchers exist. So we shall not remove the zkConnectionManager from the pool without a proper check.

Reply:
Thank you, i will understand what condition needs to be taken care before removing from the pool and come back and propose